assembly/src/assembler/mast_forest_builder.rs

```diff
@@ -347,18 +337,17 @@ impl MastForestBuilder {
 impl MastForestBuilder {
     /// Adds a node to the forest, and returns the [`MastNodeId`] associated with it.
     ///
-    /// If a [`MastNode`] which is equal to the current node was previously added, the previously
-    /// returned [`MastNodeId`] will be returned. This enforces this invariant that equal
-    /// [`MastNode`]s have equal [`MastNodeId`]s.
-    fn ensure_node(&mut self, node: MastNode) -> Result<MastNodeId, AssemblyError> {
-        let node_digest = node.digest();
+    /// Note adding the same [`MastNode`] twice will result in two different [`MastNodeId`]s being
+    /// returned.
+    pub fn ensure_node(&mut self, node: MastNode) -> Result<MastNodeId, AssemblyError> {
+        let node_hash = node.eq_hash();
 
-        if let Some(node_id) = self.node_id_by_hash.get(&node_digest) {
+        if let Some(node_id) = self.node_id_by_hash.get(&node_hash) {
             // node already exists in the forest; return previously assigned id
             Ok(*node_id)
         } else {
             let new_node_id = self.mast_forest.add_node(node)?;
-            self.node_id_by_hash.insert(node_digest, new_node_id);
+            self.node_id_by_hash.insert(node_hash, new_node_id);
 
             Ok(new_node_id)
         }
```

Comment:
Good point. Since `eq_hash()` was moved to `MastForestBuilder` in #1466, I added this as a failing test + the fix there ([f26e51d](https://github.com/0xPolygonMiden/miden-vm/pull/1466/commits/f26e51d59a984c153641ab3287ceca0d09381e7e))